core/src/main/java/net/adoptopenjdk/icedteaweb/resources/ResourceTracker.java

```diff
@@ -381,64 +364,61 @@ private Resource getResource(URL location) {
         }
     }
 
-    /**
-     * Wait for some resources.
-     *
-     * @param resources the resources to wait for
-     * @throws InterruptedException if another thread interrupted the wait
-     */
-    private void wait(Resource... resources) throws InterruptedException {
-        // save futures in list to allow parallel start of all resources
-        final List<Future<Resource>> futures = Stream.of(resources)
-                .map(ResourceHandler::new)
-                .map(ResourceHandler::putIntoCache)
-                .collect(Collectors.toList());
-
-        for (Future<Resource> future : futures) {
-            try {
-                future.get();
-            } catch (ExecutionException ignored) {
-            }
-        }
+    private Future<Resource> triggerDownloadFor(Resource resource, final Executor downloadExecutor) {
+        return new ResourceHandler(resource).putIntoCache(downloadExecutor);
     }
 
     /**
      * Wait for some resources.
      *
      * @param resources the resources to wait for
-     * @param timeout   the timeout, or {@code 0} to wait until completed
      * @return {@code true} if the resources were downloaded or had errors,
      * {@code false} if the timeout was reached
      * @throws InterruptedException if another thread interrupted the wait
      */
-    private boolean wait(Resource[] resources, long timeout, TimeUnit timeUnit) throws InterruptedException {
-        if (timeout <= 0) {
-            throw new IllegalArgumentException("Timout must be bigger than 0");
+    private void waitForCompletion(Resource... resources) {
+        if (resources == null || resources.length == 0) {
+            return;
         }
-        long startTime = System.nanoTime();
-        long nanoTimeout = timeUnit.toNanos(timeout);
 
-        // save futures in list to allow parallel start of all resources
-        final List<Future<Resource>> futures = Stream.of(resources)
-                .map(ResourceHandler::new)
-                .map(ResourceHandler::putIntoCache)
-                .collect(Collectors.toList());
+        final int configuredThreadCount = Integer.parseInt(JNLPRuntime.getConfiguration().getProperty(ConfigurationConstants.KEY_PARALLEL_RESOURCE_DOWNLOAD_COUNT));
+        final int threadCount = Math.min(configuredThreadCount, resources.length);
+        final ExecutorService downloadExecutor = Executors.newFixedThreadPool(threadCount, new DaemonThreadFactory());
+        try {
+            Arrays.asList(resources).stream()
+                    .map(r -> triggerDownloadFor(r, downloadExecutor))
```

Comment:
Do we need to collect **all** the futures in a list before we start calling `get()` on them?

Reply:
What do you mean?

Reply:
The `triggerDownloadFor` returns a Future. This happens very fast (not blocking).
But on the next line the forEach will wait for the first Future to complete before the stream will process the next Rrsource.
Or at least that is how I think streams behave.